book/src/proposals/abi-management.md

```diff
@@ -0,0 +1,434 @@
+
+```Rust
+pub fn serializing_with_fixture(|write_stream, revision| {
+  // serialize here according to revision
+  Ok(...)/Err(...)
+});// internally asserts against binary in the fixture for equality or assert status.
+
+pub fn deserializing_with_fixture(normal_result, |read_stream, revision| {
+  // deserialize and post-process according to revision
+  Ok(...)/Err(...)
+});// internally asserts against given result for equality or assert status.
+```
+
+## Assertion steps 
+
+### for backward compatibility
+
+We can make sure we can serialize/deserialize for given change in this way:
+
+1. Run deserialize test with fixture of REV-1, REV-2, ..., REV0
+2. Run serialize test with fixture of REV-1, REV-2, ..., REV0
+
+### for forward compatibility
+
+We can make sure mixed versions of cluster doesn't break for given change in
+this way:
+
+1. Only backport the newer test fixtures without application code changes
+2. Run deserialize test with fixture of REV+1, REV+2, ... REV+LATEST
+3. Run serialize test with fixture of REV+1, REV+2, ... REV+LATEST
+
+
+# Developer's work flow
+
+We will end up accumulating increasing number of technical debt over time to
+support older ABI item revisions. As a general rule of thumb for implementation
+design, we always normalize the older serialized data into the runtime/internal
+representation which closely reflects the latest revision of given ABI item as
+much as possible with additional conversion logic.
+
+Because it's more expensive to maintain this ABI test than normal one, normal
+unit test is preferred for general unit testing as before. ABI unit test should
+be restricted to the bare minimal set of variation of serialization.
+
+## Change an existing compatibility item:
+
+1. Change `#[solana_abi]` `struct`s as usually
+2. Depending on its maintenance tier, decide hard-fork or normal update
+3. (For hard-fork,) decide scheduled release time based on epochs
+4. Write the code to differentiate depending on
+   [version](#how to detect versions)
+5. Generate new test fixture for the new compatibility item revision and check
+   the diff.
+
+## Create new compatibility item:
+
+1. As there is no free standing system component, this thing is actually
+   related to the other part of system in some way. Consider what compatibility
+   status and maintenance tier is expected for the relation regarding adding
+   something new.
+2. Write entirely new struct/module code with `#[solana_abi]`
+3. Write entirely new serialization/deserialization unit test code
+4. Generate fixture for the first revision and check the diff.
+
+## How to generate test fixtures for newer versions
+
+run `cargo test` with specific environment like `SOLANA_UPDATE_ABI_FIXTURES=1`.
+Then, these fixtures will be automatically updated and review with `git diff`
+
+# Implemention remarks
+
+We'll end up heavily relying on macro magics. For precedent, `ink' from the
+Parity Technologies [2], could be informational.
+
+# Considerations
```

Comment:
I want these remaining concern to be clarified or I should propose a concrete proposal for each. :)

Reply:
With scoping down of this proposal, there is no considerations!